Image Block: Add "full width" control option

aduth opened the pull request:
Related: #310
Related: #416

This pull request seeks to add a new "Full width" option to the image block controls, intended for cases where the image is to break out of any constraining container.

![full](https://cloud.githubusercontent.com/assets/1779930/25454439/6cba5d1e-2a9a-11e7-9c43-7468f8a11950.gif)

__Open questions:__

- Label name? There's some heated debate in the original issue.

__Testing instructions:__

Verify that the full width option can be toggled like other image alignment controls (#416), and that it visually breaks out of the editor container. 

__Future considerations:__

- Currently I'm using an `align-center` icon, as there's not a Dashicon yet which fits well for the intended action. @jasmussen [has committed to incorporating one](https://wordpress.slack.com/archives/C02QB2JS7/p1493236482591946?thread_ts=1493114261.780578&cid=C02QB2JS7).
- As a new option, existing themes are unlikely to support this. We'll need to consider some hooks by which a theme can define its support and conditionally show the control.

aduth committed:
Image Block: Add "full width" control option

BE-Webdesign commented on blocks/library/image/style.scss:
```diff
@@ -15,6 +15,11 @@
+	&[data-align="full"] {
+		width: 90%;
+		max-width: none;
+	}
```
Should we center the image as well?

```css
&[data-align="full"] {
    img {
        margin: 0 auto;
    }
}
```

Not a huge fan of over nesting selectors but YOLO!

BE-Webdesign commented:
The formatting and application of the attribute works 👌, however when the image goes full width on a larger screen it is not centered, like I would expect it to be.  Not sure what the expected behavior should be.  Other than that, it looks good to me.

Also totally 👍 on data-align=full, I saw somewhere else where it was brought up that full-width was not an alignment.  Maybe that is true, but I am pretty sure most people will understand and this works a lot cleaner than having all sorts of names.

BE-Webdesign commented on editor/modes/visual-editor/index.js:
```diff
@@ -16,7 +16,9 @@ function VisualEditor( { blocks } ) {
-			<Inserter />
+			<div className="editor-visual-editor__container">
+				<Inserter />
+			</div>
```
Any context to this change for this PR?

aduth replied:
>Any context to this change for this PR?

To preserve the common max-width of 700px across blocks while still allowing blocks to escape out of it, the responsibility of assigning the max width was moved from the listing to the block itself. Without the changes here, the inserter would be shown to the far left of the editor canvas since it's not a true block, but should respect aligned with of the blocks (effected with the container).

BE-Webdesign replied:
Ah makes perfect sense.

iseulde commented on blocks/library/image/style.scss:
```diff
@@ -15,6 +15,11 @@
 	}
+
+	&[data-align="full"] {
+		width: 90%;
```
What's 90% here?

BE-Webdesign replied:
I think it is trying to go full-width but not too full-width. 100% would probably make more sense though.

aduth replied:
It was a number pulled out of the air 😄  But yes, specifically it's a value close to 100% while still allowing space for the block mover.

jasmussen commented:
Added a new full width icon in https://github.com/WordPress/gutenberg/pull/521

![7a0f4a74-2b35-11e7-8be7-8f8d832c8fe6](https://cloud.githubusercontent.com/assets/1204802/25475072/e029b31c-2b35-11e7-91b0-ce857fd0a4c2.png)

`align-full-width`

mtias commented:
Maybe just `wide` as attribute name?

aduth commented:
>however when the image goes full width on a larger screen it is not centered, like I would expect it to be.

Good catch on this @BE-Webdesign . I guess there's a question of: Do we want to force the image to take up the [full width of the canvas](https://cloud.githubusercontent.com/assets/1779930/25488718/bd3e8760-2b35-11e7-879d-331175165c9e.png)? Its [natural width](https://cloud.githubusercontent.com/assets/1779930/25488719/bd489d40-2b35-11e7-9418-29b9b10e3ca2.png) (with `margin: 0 auto`)? Some arbitrary "expanded" width (like `max-width: 900px` instead of `700px`)?

BE-Webdesign commented:
I would say natural width to prevent images from overly distorting.  If you want an image to fill up more space use a bigger image.  On super hi res screens, if we go full width of the canvas, images will look really silly compared to the text content.  The arbitrary "expanded" might work the best, I would go with 1130px for the max-width for a nice ratio between the two.

jasmussen commented:
> Good catch on this @BE-Webdesign . I guess there's a question of: Do we want to force the image to take up the full width of the canvas? Its natural width (with margin: 0 auto)? Some arbitrary "expanded" width (like max-width: 900px instead of 700px)?

I meant to respond to this previously, but clearly missed out on this one too.

Full-width, née "full bleed", is close to my heart. I feel strongly that it should be _truly_ full width. Edge to edge. Definitely in the editor, and depending on technical implementations, also in the theme itself. 

Try viewing [the example post](http://moc.co/sandbox/example-post/), and scroll down to this image:

![screen shot 2017-04-27 at 16 53 59](https://cloud.githubusercontent.com/assets/1204802/25489445/26f000ec-2b6a-11e7-9572-aab0725922cb.png)

That's the look that we're going for with this one. Big and bold. If we are too concerned about it being mega huge, we could add another option which is could be the "wide", but I still think there's a case for going edge to edge. 

That being said, there are some technical challenges. The UI prototype used this trick to go edge to edge, despite any max-widths applied to the main content colum:

```
.align-full-bleed {
    margin-left: calc(50% - 50vw);
    width: 100vw;
}
```

The problem with that one is that the `vw` unit does not account for scrollbars, which means any browser that doesn't have fancy mac auto hiding scrollbars will also get a horizontal scrollbar. In my mind that is a blocker for this method of full width images. The example post referenced above explicitly avoids this implementation by not having a max-width on the main column. 

This suggests that we may want to leave it up to the theme itself to decide what to do, and simply apply a CSS class to the image.

aduth commented:
>This suggests that we may want to leave it up to the theme itself to decide what to do, and simply apply a CSS class to the image.

Yeah, I think it really depends on how the theme authors would use this attribute. The class already exists with the changes here, it's just deciding how to display it in the editor. The problem with the prototype version is that there's no way to move the block up and down, hence the 90% here.

jasmussen commented:
> Yeah, I think it really depends on how the theme authors would use this attribute. The class already exists with the changes here, it's just deciding how to display it in the editor. The problem with the prototype version is that there's no way to move the block up and down, hence the 90% here.

Yep. We had some ideas for this, and I'm still thinking about that. Right at this moment, I think it's okay that you can't move it when full bleed (because the controls are off screen). If you need to move it, you can temporarily pick a different alignment, or move the blocks before and after, or use drag and drop. 

However if in practice we find that this is annoying, let's revisit and perhaps we can move the position of the mover for this look.

aduth committed:
Remove unused transparent block border

Might have been duplicated from :before pseudo-element hover/focus
bounds border

aduth commented:
Rebased to address feedback and incorporate the new icon.

But there's a hiccup now in that floated images no longer respect the block maximum width. Will have to poke at this a while longer to see if there's another way to "escape" the container.

jasmussen commented:
I've noted down a few articles on how to escape the container, perhaps worth looking at: https://css-tricks.com/full-width-containers-limited-width-parents/

There was also this one, which I thought was particularly promising: https://css-tricks.com/hassle-free-full-bleed-not/

aduth committed:
Use margin calculation for full bleed image offsets

aduth commented:
@jasmussen Thanks for the links. That first one was particularly great in breaking down considerations; for us, the "non-% Parent Width" case. Inevitably this includes the viewport width unit which, as you've mentioned, doesn't account for the presence of the sidebar. I still went down this path in a2d8372, using `overflow: hidden;` on the editor container to prevent the horizontal scroll. This also means there's likely ~25px of the image clipped when a scrollbar exists. Also, as can be seen in the CSS, managing offsets for varying sidebar collapsed / folded / hidden states is quite cumbersome. Ugly as it is, it should all be accounted for now.

To the other link: The issue is that if we move the burden of managing max-width and centering to the block itself, it becomes equally difficult to come up with a solution for the floated images to be aligned to the invisible left and right barriers of adjacent blocks.

jasmussen commented:
> To the other link: The issue is that if we move the burden of managing max-width and centering to the block itself, it becomes equally difficult to come up with a solution for the floated images to be aligned to the invisible left and right barriers of adjacent blocks.

Yep, I learned of this in [the example post](http://moc.co/sandbox/example-post/) with the left floated quote.

👍

aduth commented:
@jasmussen How do you feel about the solution using `vw` units, since I know you tend to have opinions on its usage 😄

jasmussen commented:
> @jasmussen How do you feel about the solution using vw units, since I know you tend to have opinions on its usage 😄

Haha yes indeed.

I personally find the unit completely useless as it presumes there are no scrollbars, and especially for Windows users, there are. However I'm also pragmatic enough to swallow that pride when presented with something that works. If it works, and works well, ship it! (And just to be clear, "works" means that there are no horizontal scrollbars, when Mac has them set to always show.)

I'm assuming themes can still choose not to use `vw` units for the frontend rendering, correct?

aduth commented:
>I'm assuming themes can still choose not to use `vw` units for the frontend rendering, correct?

Theme styling in general needs more thought, but it would be my assumption that this would be entirely up to the theme to implement based on the presence of the class, yes.
